AX: Voice Control number/name overlays don't label content in iframe

twilco opened the pull request:
#### 250a3ab845c5672c4579c6fcc406ffdf5d800c7d
<pre>
AX: Voice Control number/name overlays don&apos;t label content in iframe
<a href="https://bugs.webkit.org/show_bug.cgi?id=285188">https://bugs.webkit.org/show_bug.cgi?id=285188</a>
<a href="https://rdar.apple.com/142096466">rdar://142096466</a>

Reviewed by Joshua Hoffman.

This happened for two reasons:

  1. AccessibilityScrollView did not implement boundingBoxRect(), which AccessibilityObject::isOnScreen relied on.
     It did implement elementRect(), so implement boundingBoxRect() in terms of elementRect. In the future, we should
     consider only having one method to return an element&apos;s rect.

  2. The rect of elements within non-root scrollables (like iframes) needs to be normalized to the root view. Otherwise,
     the rect would be relative to the containing scrollable, e.g. a button at x=8 y=8 in an iframe that is x=300 y=300
     should be considered to have a position of x=308 y=308, but this wasn&apos;t happening, so we were considering the button
     to not be visible.

Two new tests added.

* LayoutTests/accessibility/mac/visible-content-search-in-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-iframe.html: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-scrolled-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-scrolled-iframe.html: Added.
* LayoutTests/resources/accessibility-helper.js:
(dumpAXSearchTraversal):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dbgInternal const):
(WebCore::AccessibilityObject::isOnScreen const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::extraDebugInfo const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::ownerDebugDescription const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:

Canonical link: <a href="https://commits.webkit.org/297816@main">https://commits.webkit.org/297816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/770dfc8142da417ea6c1c2ade0b3236201e368bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119171 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41265 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85976 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101614 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25908 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62929 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122392 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40045 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/29856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94824 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94564 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24142 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39700 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17512 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36158 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39931 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45430 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39572 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41309 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3578,37 +3584,39 @@ static int computeBestScrollOffset(int currentScrollOffset, int subfocusMin, int
+
+        IntRect innerRect;
+        if (inner->isScrollView())
+            innerRect = snappedIntRect(inner->parentObject()->boundingBoxRect());
```
Is the ScrollView returning the bounding box rect of the content rather than the view?

Two things:

1. Probably need a null check - we know that a scroll view might temporarily return nullptr for parentObject()
2. I wonder if it'd be cleaner to just introduce a new method on AccessibilityObject like boundingBoxRectOnScreen and have AccessibilityScrollView override it, rather than having a special case in this loop.

twilco replied:
For what it's worth, this isn't new code — I just extracted it out of the existing ternary. I don't know what the author's intention was when deciding to use `inner->parentObject()`'s bounding box if `isScrollView()`. But what's interesting is that if I remove this special-case, no tests fail.

Perhaps this special-case was to workaround AccessibilityScrollView not implementing boundingBoxRect? However, I still think implementing that is essential, as even if we kept this special parentObject case for computing `innerRect`, not having `boundingBoxRect` implemented for `AccessibilityScrollView` causes issues when computing `outerRect` for `iframe`s (it's the reason this bug happens).

So I'm inclined to solve this by removing the scrollView-parentObject special case, and have done so in the latest commit.

But let me know if you have other ideas on approaching this, e.g. if you feel strongly that we should explore that other approach you mentioned. IIRC the gist of that was comparing the rects of the object-under-test and the visible page rect, where each rect is computed through the same means we take for serving NSAccessibilityPosition, rather than the manual boundingBoxRect approach this method currently takes.

twilco replied:
Also I'm sorry, I realized I never answered your question:

> Is the ScrollView returning the bounding box rect of the content rather than the view?

`AccessibilityScrollView::boundingBoxRect()` / `AccessibilityScrollView::elementRect` do not return the rect of the content. Here is some logging I took while running a layout test — we have a 4000px-height div inside the iframe, but the scrollview is only 150px-height.

```
outer {role: ScrollArea, ID 83 (scrollview 0x10801b400, owned by: RenderIFrame 0x108214700 IFRAME 0x108194460 id='iframe')}
>>> rect (107,10) width=300 height=150

<iframe id="iframe" onload="runTest()" srcdoc="
    <div style='height: 4000px;'></div>
    <button id='inside-iframe-button'>Inside iframe</button>
">
</iframe>
```

It's implemented like so, so `frameRectShrunkByInset` must not account for content size.

```cpp
LayoutRect AccessibilityScrollView::elementRect() const
{
    RefPtr scrollView = currentScrollView();
    return scrollView ? scrollView->frameRectShrunkByInset() : LayoutRect();
}
```

minorninth replied:
Thanks for investigating and explaining!

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3578,37 +3584,39 @@ static int computeBestScrollOffset(int currentScrollOffset, int subfocusMin, int
+        else {
+            innerRect = snappedIntRect(inner->boundingBoxRect());
+            if (RefPtr scrollView = !outer->isRoot() ? outer->scrollView() : nullptr)
+                innerRect = scrollView->contentsToRootView(innerRect);
```
Can you do this without the "if"? Can contentsToRootView just be a no-op for the root view?

twilco replied:
> Can you do this without the "if"? Can contentsToRootView just be a no-op for the root view?

I played around with this idea, and I think the answer is no unfortunately. Despite the method name implying no work should be done if this scroll view is the root, it will unconditionally call `ScrollView::contentsToView` on the passed in rect, which results in the wrong behavior (failing layout tests, Voice Control missing entire swaths of content in Safari on https://en.wikipedia.org/wiki/Main_Page after scrolling the page).

twilco committed:
AX: Voice Control number/name overlays don't label content in iframe
https://bugs.webkit.org/show_bug.cgi?id=285188
rdar://142096466

Reviewed by Joshua Hoffman.

This happened for two reasons:

  1. AccessibilityScrollView did not implement boundingBoxRect(), which AccessibilityObject::isOnScreen relied on.
     It did implement elementRect(), so implement boundingBoxRect() in terms of elementRect. In the future, we should
     consider only having one method to return an element's rect.

  2. The rect of elements within non-root scrollables (like iframes) needs to be normalized to the root view. Otherwise,
     the rect would be relative to the containing scrollable, e.g. a button at x=8 y=8 in an iframe that is x=300 y=300
     should be considered to have a position of x=308 y=308, but this wasn't happening, so we were considering the button
     to not be visible.

Two new tests added.

* LayoutTests/accessibility/mac/visible-content-search-in-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-iframe.html: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-scrolled-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/visible-content-search-in-scrolled-iframe.html: Added.
* LayoutTests/resources/accessibility-helper.js:
(dumpAXSearchTraversal):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dbgInternal const):
(WebCore::AccessibilityObject::isOnScreen const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::extraDebugInfo const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::ownerDebugDescription const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:

Canonical link: https://commits.webkit.org/297816@main
